airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/SnowflakeGcsStagingSqlOperations.java

```diff
@@ -0,0 +1,197 @@
+/*
+ * Copyright (c) 2021 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.integrations.destination.snowflake;
+
+import static io.airbyte.integrations.destination.snowflake.SnowflakeInternalStagingSqlOperations.UPLOAD_RETRY_LIMIT;
+
+import com.google.auth.oauth2.GoogleCredentials;
+import com.google.cloud.storage.BlobId;
+import com.google.cloud.storage.BlobInfo;
+import com.google.cloud.storage.BucketInfo;
+import com.google.cloud.storage.Storage;
+import com.google.cloud.storage.StorageOptions;
+import io.airbyte.commons.lang.Exceptions;
+import io.airbyte.commons.string.Strings;
+import io.airbyte.db.jdbc.JdbcDatabase;
+import io.airbyte.integrations.base.sentry.AirbyteSentry;
+import io.airbyte.integrations.destination.NamingConventionTransformer;
+import io.airbyte.integrations.destination.jdbc.copy.gcs.GcsConfig;
+import io.airbyte.integrations.destination.record_buffer.SerializableBuffer;
+import io.airbyte.integrations.destination.staging.StagingOperations;
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.channels.Channels;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.UUID;
+import org.joda.time.DateTime;
+
+public class SnowflakeGcsStagingSqlOperations extends SnowflakeSqlOperations implements StagingOperations {
+
+  private final NamingConventionTransformer nameTransformer;
+  private final Storage storageClient;
+  private final GcsConfig gcsConfig;
+  private final Set<String> fullObjectKeys = new HashSet<>();
+
+  public SnowflakeGcsStagingSqlOperations(NamingConventionTransformer nameTransformer, GcsConfig gcsConfig) {
+    this.nameTransformer = nameTransformer;
+    this.gcsConfig = gcsConfig;
+    this.storageClient = getStorageClient(gcsConfig);
+  }
+
+  private Storage getStorageClient(GcsConfig gcsConfig) {
+    try {
+      final InputStream credentialsInputStream = new ByteArrayInputStream(gcsConfig.getCredentialsJson().getBytes(StandardCharsets.UTF_8));
+      final GoogleCredentials credentials = GoogleCredentials.fromStream(credentialsInputStream);
+      return StorageOptions.newBuilder()
+          .setCredentials(credentials)
+          .setProjectId(gcsConfig.getProjectId())
+          .build()
+          .getService();
+    } catch (Exception e) {
+      throw new RuntimeException(e);
+    }
+
+  }
```

Comment:
Great that you are using GCS client instead of S3. Some of the logics here can be moved to `GcsStorageOperations` so that all the GCS related destination can use this native client. But this can be done in a separate PR.

Reply:
@tuliren created new issue https://github.com/airbytehq/airbyte/issues/13313